rewrite with ts and router in mind

giogonzo opened the pull request.

francescogior commented on src/connect.tsx:
```diff
@@ -0,0 +1,163 @@
+      const shouldUpdateState = filterValid ? (v: S) => {
+        const invalid = decl.reduce((acc, k) => {
+          return !(stateType.meta.props[k] as t.Type<any>).is(v[k]) ? [...acc, k] : acc; // TODO(typo)
+        }, []);
```
you may use `filter` instead

giogonzo replied:
👍

francescogior commented on src/transition.ts:
```diff
@@ -0,0 +1,54 @@
+    const newState = stateType(
+      // TODO(typo): double check, it was:
+      // transitionReducer(shouldReplace ? { ...patch } : { ...state, ...patch })
+      transitionReducer(shouldReplace ? Object.assign({}, patch) : Object.assign({}, state, patch))
```
TS ha troubles with spread objects operator?

giogonzo replied:
Not with spread types in general, but `[ts] Spread types may only be created from object types`.
I think this is just because the type parameter `S` doesn't extend `{}` directly here.

Added a comment to clarify

FrancescoCioria commented on package.json:
```diff
@@ -23,16 +24,49 @@
+      "json",
+      "jsx"
+    ],
+    "testURL": "http://test.example.com"
```
?

giogonzo replied:
I was testing with browser history and wanted to have a valid path to be able to access the browser location from jsdom. I ended up manually mocking since I realized jest (and jsdom) don't support this at all. Will remove

FrancescoCioria commented on package.json:
```diff
@@ -23,16 +24,49 @@
+    "react-dom": "^15",
+    "react-test-renderer": "^15.6.1",
     "sinon": "^1.17.4",
     "smooth-release": "^6.0.0",
```
let's update this too?

giogonzo replied:
👍

FrancescoCioria commented on src/connect.tsx:
```diff
@@ -0,0 +1,163 @@
+  state: PropTypes.object.isRequired
+};
+
+function defaultGetNewState<S extends ST, Decl extends string>(ty: Decl[]) {
```
what does `ty` stand for?

giogonzo replied:
will update to something clearer (this is old code, just transpiled)

FrancescoCioria commented on src/connect.tsx:
```diff
@@ -0,0 +1,163 @@
+          }
+        }
+
+        shouldComponentUpdate(nextProps: ConnectedComponentProps<P, Decl>, nextState: StateT<S, Decl>) {
```
remove in favor of `PureComponent`?

giogonzo replied:
not super easy to do because it is based on a config parameter (`pure`).. I think we'd need two different declarations. I'll leave it as is for now

FrancescoCioria commented on src/index.ts:
```diff
@@ -0,0 +1,34 @@
+
+  const stateType = t.interface(
+    {
+      ___k: t.maybe(t.Number),
```
remove?

FrancescoCioria commented on src/mkContextWrapper.tsx:
```diff
@@ -0,0 +1,15 @@
+import * as React from 'react';
+
+export default function mkContextWrapper(ctx: {}, ctxTypes: {}): React.ComponentType<{ children: () => JSX.Element }> {
```
consider using `{ [k: string]: PropType }` instead of `{}`

FrancescoCioria replied:
even better, if we don't export `mkContextWrapper` `ctx` can perfectly typed as it's always `{ transition, state }`

FrancescoCioria commented on src/shallowEqual.ts:
```diff
@@ -0,0 +1,29 @@
+type Obj = null | undefined | { [k: string]: any };
```
this file may become useless if we use `PureComponent`

FrancescoCioria commented on src/transition.ts:
```diff
@@ -0,0 +1,54 @@
+      transitionReducer(shouldReplace ? Object.assign({}, patch) : Object.assign({}, state, patch))
+    );
+
+    const stateChanged = !shallowEqual(state, newState);
```
oh right, we're using it here too, nevermind

giogonzo commented:
@FrancescoCioria @francescogior added parsing (ported from #40), added tests and fixed more issues,  addressed all review comments

FrancescoCioria commented on src/browser.ts:
```diff
@@ -11,7 +11,7 @@ export default function(history: History = createBrowserHistory()) {
 
   function syncToBrowser(s: BrowserState, push: boolean = true): void {
-    const pathname = `/${s.view}`;
+    const pathname = `/${s.view || ''}`;
```
then update BrowserState to have `view` optional? (right now it is `view: string`)

giogonzo replied:
no I forgot this. It's responsibility of the app to enforce an always-there `view` key (and this is reflected in the static state type and, as a consequence, in `BrowserState` too). So no need for this default here.
I initially put the default there because on alinity pro we had:
```js
initialState = { /* ... no view key ... */ }
// and
transitionReducer = state => {
  if (state.view === 'main') { 
    state.view = defaultView
  }
  // ...
  return state;
}
```
I updated the check in `transitionReducer` to be `if (!state.view)` and it works 👍

FrancescoCioria commented on src/browser.ts:
```diff
@@ -16,9 +16,9 @@ export default function(history: History = createBrowserHistory()) {
-  function parseBrowserState(location: typeof history.location): BrowserState {
-    const view = trim(location.pathname, ' /');
-    const s = qs.parse(location.search, { ignoreQueryPrefix: true });
+  function parseBrowserState(location?: typeof history.location): BrowserState {
```
when can `location` be missing?

giogonzo replied:
just because of tests I think.. added a `TODO` with clarifying comment

giogonzo commented:
@FrancescoCioria @francescogior also see https://github.omnilab.our.buildo.io/buildo/alinity-pro/pull/6406

FrancescoCioria commented on src/index.ts:
```diff
@@ -18,13 +18,7 @@ export default function<S extends StateT>(
-    },
-    { strict: true, name: _stateType.meta.name }
-  ) as t.Interface<S>;
+  const stateType = t.interface(_stateType.meta.props, { strict: true, name: _stateType.meta.name }) as t.Interface<S>;
```
at this point `stateType` is the same as `_stateType`, we could remove this logic completely

giogonzo replied:
👍

giogonzo commented:
@FrancescoCioria @francescogior again, addressed all comments
The last 8 commits, other than addressing @FrancescoCioria comments, are adding 2 tests for "user interacting with browser history" and making them pass. The code is still a bit messy (last commit)
Tested on alinity pro 🎉
